server/src/future.rs

```diff
@@ -152,92 +146,89 @@ where
 
-impl Drop for StopMonitor {
-	fn drop(&mut self) {
-		if Arc::strong_count(&self.0) == 1 {
-			self.0.waker.wake();
-		}
-	}
-}
-
-impl StopMonitor {
-	pub(crate) fn new() -> Self {
-		StopMonitor(Arc::new(MonitorInner { shutdown_requested: AtomicBool::new(false), waker: AtomicWaker::new() }))
+impl StopHandle {
+	pub(crate) fn new(rx: watch::Receiver<()>) -> Self {
+		Self(rx)
 	}
 
 	pub(crate) fn shutdown_requested(&self) -> bool {
-		// We expect this method to be polled frequently, skipping an iteration isn't problematic, so relaxed
-		// ordering is optimal.
-		self.0.shutdown_requested.load(Ordering::Relaxed)
+		// if a message has been seen, it means that `stop` has been called.
+		self.0.has_changed().unwrap_or(true)
 	}
 
-	pub(crate) fn handle(&self) -> ServerHandle {
-		ServerHandle(Arc::downgrade(&self.0))
+	pub(crate) async fn shutdown(&mut self) {
+		// Err(_) implies that the `sender` has been dropped;
+		// Ok(_) implies that `stop` has been called,
+		let _ = self.0.changed().await;
 	}
 }
 
-/// Handle that is able to stop the running server or wait for it to finish
-/// its execution.
-#[derive(Debug, Clone)]
-pub struct ServerHandle(Weak<MonitorInner>);
+/// Server handle.
+///
+/// When all server handles has been `dropped` or `stop` has been called
+/// the server will be stopped.
+#[derive(Debug)]
+pub struct ServerHandle {
+	inner: watch::Sender<()>,
+	heartbeat: Interval,
+}
 
 impl ServerHandle {
-	/// Requests server to stop. Returns an error if server was already stopped.
+	/// Create a new server handle.
+	pub fn new(tx: watch::Sender<()>) -> Self {
+		let mut heartbeat = time::interval(STOP_MONITOR_POLLING_INTERVAL);
+		heartbeat.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+
+		Self { inner: tx, heartbeat }
+	}
+
+	/// Stop the server
 	///
-	/// Returns a future that can be awaited for when the server shuts down.
-	pub fn stop(self) -> Result<ShutdownWaiter, Error> {
-		if let Some(arc) = Weak::upgrade(&self.0) {
-			// We proceed only if the previous value of the flag was `false`
-			if !arc.shutdown_requested.swap(true, Ordering::Relaxed) {
-				return Ok(ShutdownWaiter(self.0));
-			}
-		}
-		Err(Error::AlreadyStopped)
+	/// This may be called several times but has no effect.
+	pub fn stop(self) -> Result<impl Future<Output = ()>, Error> {
+		self.inner.send(()).map_err(|_| Error::AlreadyStopped)?;
+		Ok(self)
+	}
+
+	/// Check if the server has been stopped.>
+	pub fn is_stopped(&self) -> bool {
+		self.inner.is_closed()
 	}
 }
 
 impl Future for ServerHandle {
 	type Output = ();
 
-	fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-		let mut shutdown_waiter = ShutdownWaiter(self.0.clone());
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		// We don't care about the ticks of the heartbeat, it's here only
+		// to periodically wake the `Waker` on `cx`.
+		let _ = self.heartbeat.poll_tick(cx);
```

Comment:
I'm not sure if this is the correct thing to but the future wasn't woken up otherwise.

Reply:
Another alternative to the heartbeat might be something similar to

```rust
		if self.inner.is_closed() {
			Poll::Ready(())
		} else {
			cx.waker().wake_by_ref();
			Poll::Pending
		}
```

But this will generally start `polling` and steal some time from the scheduler.

Reply:
right, I like that better